src/panels/config/js/trigger/device.tsx

```diff
@@ -2,29 +2,58 @@ import { h, Component } from "preact";
 
 import "../../../../components/device/ha-device-picker";
 import "../../../../components/device/ha-device-trigger-picker";
+import "../../../../components/ha-form";
+import "../../../../components/paper-time-input";
+
+import { fetchDeviceTriggerCapabilities } from "../../../../data/device_automation";
 
 export default class DeviceTrigger extends Component<any, any> {
+  private hass;
   constructor() {
     super();
+
     this.devicePicked = this.devicePicked.bind(this);
     this.deviceTriggerPicked = this.deviceTriggerPicked.bind(this);
-    this.state = { device_id: undefined };
+    this._extraFieldsChanged = this._extraFieldsChanged.bind(this);
+    this.state = { device_id: undefined, capabilities: undefined };
   }
 
   public devicePicked(ev) {
-    this.setState({ device_id: ev.target.value });
+    this.setState({ ...this.state, device_id: ev.target.value });
   }
 
-  public deviceTriggerPicked(ev) {
+  public async deviceTriggerPicked(ev) {
     const deviceTrigger = ev.target.value;
+    const capabilities = deviceTrigger.domain
+      ? await fetchDeviceTriggerCapabilities(this.hass, deviceTrigger)
+      : null;
+    this.setState({ ...this.state, capabilities });
     this.props.onChange(this.props.index, deviceTrigger);
   }
 
   /* eslint-disable camelcase */
   public render({ trigger, hass }, { device_id }) {
+    this.hass = hass;
+    if (!this.state.capabilities && trigger.domain) {
+      fetchDeviceTriggerCapabilities(this.hass, trigger).then(
+        (capabilities) => {
+          this.setState({ ...this.state, capabilities });
+        }
+      );
+    }
```

Comment:
`fetchDeviceTriggerCapabilities` is called when device trigger is picked, but if we're editing an existing automation that won't happen. The purpose of this is just to load the trigger capabilities once. 
OK to take care of it here, or should it be done elsewhere, for example in `componentDidMount`?

Reply:
The render method should never have side effects. Not in Preact nor Lit. It is possible that this method can get called multiple times before the capabilities resolve, and you will do many requests.

Instead, do it inside  [`componentDidUpdate`](https://preactjs.com/guide/v8/api-reference/#lifecycle-methods).  (similar to `updated` in Lit)

Reply:
OK, moved out of render but to `componentDidMount`, `componentDidUpdate` was not called after first render.